Add alias as an official Azure CLI extension

chewong opened the pull request:
Add alias as an official Azure CLI extension. Also fixed `src/index.json` as it contains a mix of spaces and tabs, and converted to 4-space indented.

@troydai and @derekbekoe, could you review this PR? Thanks!

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `./scripts/ci/test_static.sh` locally? (`pip install pylint flake8` required)
- [x] Have you run `python scripts/ci/test_integration.py -q` locally?

azuresdkci commented:
If this PR is for a new extension or change to an existing extension, use the following to try out the changes in this PR:
```
docker run -it microsoft/azure-cli:latest
export EXT=<NAME>
pip install --upgrade --target ~/.azure/cliextensions/$EXT "git+https://github.com/chewong/azure-cli-extensions.git@alias-extension#subdirectory=src/$EXT&egg=$EXT"
```

chewong commented:
Sorry for accidentally tagging @yugangw-msft as a code reviewer

derekbekoe commented:
@chewong When CI is green, let me know

troydai commented on src/alias-preview/setup.py:
```diff
@@ -0,0 +1,45 @@
+
+DEPENDENCIES = [
+    'ddt==1.1.1',
+    'configparser==3.5.0'
```
use `~=`

chewong commented on src/alias-preview/azext_alias/telemetry/telemetry.py:
```diff
@@ -0,0 +1,128 @@
+
+
+# pylint: disable=too-many-instance-attributes
+class AliasExtensionTelemetrySession(object):
```
This is a helper class similar `telemetry.py` in the core that collects telemetry data from the extension. A `AliasExtensionTelemetrySession` represents the duration of the alias transformation.`alias.py` rely on this class to temporarily store telemetry data and send it to the core once the alias transformation is done.

chewong commented:
@derekbekoe, the CI (finally) turns green. Basically, the dependency of alias (`ddt`) was installed in the temp folder, but we were running the tests against the extension folder, instead of the temp folder, that's why the tests cannot find the `ddt` module initially. I tried different methods of solving this problem and found that the simplest solution is to include `ddt.py` in the alias src folder. Please let me know if you have any concern about this approach.

derekbekoe commented on src/alias/azext_alias/__init__.py:
```diff
@@ -0,0 +1,60 @@
+from azure.cli.core.commands.events import EVENT_INVOKER_PRE_CMD_TBL_TRUNCATE
+from azext_alias.alias import AliasManager
+from azext_alias._const import DEBUG_MSG_WITH_TIMING
+from azext_alias.telemetry import telemetry
```
Move an many imports here to the actual methods that use them for perf. reasons.

I think only these ones would need to stay at the top.
```
from knack.log import get_logger
from azure.cli.core import AzCommandsLoader
```

derekbekoe commented on src/alias/azext_alias/__init__.py:
```diff
@@ -0,0 +1,60 @@
+from azext_alias.telemetry import telemetry
+
+logger = get_logger(__name__)
+VERSION = '0.1.0'
```
I don't like having the same version no. defined both here and in setup.py.

derekbekoe replied:
Consider doing something like what Python SDK does at https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt-cdn/setup.py#L45.

Alternatively, call `from azure.cli.core.extension import get_extension`. Then call it with the extension name then you can get your version that way.

derekbekoe commented on src/alias/azext_alias/__init__.py:
```diff
@@ -0,0 +1,60 @@
+    def load_command_table(self, _):  # pylint: disable=no-self-use
+        return {}
+
+    def load_arguments(self, _):
```
You didn't need the pylint disable for no-self-use here? Interesting.

derekbekoe commented on src/alias/azext_alias/_const.py:
```diff
@@ -0,0 +1,27 @@
+PLACEHOLDER_REGEX = r'\s+{\d+}'
+
+INCONSISTENT_INDEXING_ERROR = (
+    'azure-cli-alias-extension: Placeholder indexing should be zero-indexed, '
```
Hasn't the name of the extension changed?
Same below.

derekbekoe commented on src/alias/azext_alias/alias.py:
```diff
@@ -0,0 +1,322 @@
+import re
+import hashlib
+import json
+from configparser import ConfigParser
```
Would prefer you use `from six.moves import configparser`.
See https://pythonhosted.org/six/.

The configparser library you used hasn't been updated since 2016 and not sure you really need it either.

chewong replied:
Just remember why i didn't use `six.moves.configparser`: in python 2, `six.moves.configparser` is not able to detect duplicated sections and options in the config file. That means that users can create duplicated aliases or duplicated command section within the same alias without raising an error.

derekbekoe replied:
What about this?
```
When strict is True (the default), the parser won’t allow for any section or option duplicates while reading from a single source (file, string or dictionary), raising DuplicateSectionError or DuplicateOptionError. 
https://docs.python.org/3/library/configparser.html#configparser.ConfigParser
```

It would only work on Python 3 since I couldn't find that same note on Python 2 but I don't think it's worth adding a 2 year old extension just for that.
If you have duplicate sections, then I assume just the last one would be registered. That seems ok to me.

derekbekoe commented on src/alias/azext_alias/alias.py:
```diff
@@ -0,0 +1,322 @@
+        open_mode = 'r+' if os.path.exists(GLOBAL_COLLIDED_ALIAS_PATH) else 'w+'
+        with open(GLOBAL_COLLIDED_ALIAS_PATH, open_mode) as collided_alias_file:
+            collided_alias_str = collided_alias_file.read()
+            self.collided_alias = json.loads(collided_alias_str if collided_alias_str else '{}')
```
How do you handle invalid json in `collided_alias_str`?

derekbekoe commented on src/alias/azext_alias/alias.py:
```diff
@@ -0,0 +1,322 @@
+                    if word not in self.collided_alias:
+                        self.collided_alias[word] = []
+                    self.collided_alias[word].append(level)
+        telemetry.set_collided_aliases(list(self.collided_alias.keys()))
```
How is this useful for telemetry?

derekbekoe commented on src/alias/azext_alias/telemetry/telemetry.py:
```diff
@@ -0,0 +1,127 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
```
Why not just have `src/alias/azext_alias/telemetry.py`?
Seems like you only have one file under `src/alias/azext_alias/telemetry`.

derekbekoe commented on src/alias/azext_alias/telemetry/telemetry.py:
```diff
@@ -0,0 +1,127 @@
+from knack.log import get_logger
+
+from azure.cli.core import telemetry as telemetry_core
+from azure.cli.core.telemetry import _remove_cmd_chars, _remove_symbols, _get_stack_trace
```
Why are we importing private methods?
@troydai will we make them public in the future?

chewong replied:
I'll make a copy of these private methods in `telemetry.py` these private methods import.

derekbekoe commented on src/alias/azext_alias/telemetry/telemetry.py:
```diff
@@ -0,0 +1,127 @@
+from azure.cli.core._environment import get_config_dir
+import azext_alias
+
+EXTENSION_NAME = 'azure-cli-alias-extension'
```
We changed the name right?

derekbekoe commented on src/alias/azext_alias/telemetry/telemetry.py:
```diff
@@ -0,0 +1,127 @@
+from azure.cli.core import telemetry as telemetry_core
+from azure.cli.core.telemetry import _remove_cmd_chars, _remove_symbols, _get_stack_trace
+from azure.cli.core._environment import get_config_dir
+import azext_alias
```
What's this import for?

derekbekoe commented on src/alias/azext_alias/telemetry/telemetry.py:
```diff
@@ -0,0 +1,127 @@
+
+    @classmethod
+    def get_exception_message(cls, exception):
+        exception_message = str(exception).replace(get_config_dir(), '.azure')
```
You want to hide the actual users full path to .azure?

derekbekoe commented on src/alias/setup.py:
```diff
@@ -0,0 +1,45 @@
+]
+
+DEPENDENCIES = [
+    'ddt~=1.1.1',
```
ddt is only used for tests. Do we want to ship it in the product as well?

derekbekoe commented on src/alias/setup.py:
```diff
@@ -0,0 +1,45 @@
+    license='MIT',
+    author='Ernest Wong',
+    author_email='t-chwong@microsoft.com',
+    url='https://github.com/chewong/azure-cli-alias-extension',
```
Change this URL to this repo.

derekbekoe commented on src/alias/azext_alias/__init__.py:
```diff
@@ -0,0 +1,60 @@
+
+    def __init__(self, cli_ctx=None):
+        super(AliasExtensionLoader, self).__init__(cli_ctx=cli_ctx,
+                                                   min_profile='2017-03-10-profile',
```
@chewong You would probably want to remove this.
`min_profile='2017-03-10-profile'` is used so the extension isn't shown in stack profile which wouldn't be a problem here.

derekbekoe commented on .github/CODEOWNERS:
```diff
@@ -15,3 +15,5 @@
 
 /src/managementgroups/ @rajshah11
+
+/src/alias/ @chewong
```
I think so!
https://help.github.com/articles/about-codeowners/

chewong commented:
Here are some changes I made to the PR according to the comments:
- Added @troydai as a code owner
- Removed `ddt` and `configparser` dependencies
- Used `six.moves.configparser` instead of `configparser`
- Moved `ddt.py` to `azext_alias/tests` so it won't be included when installing the wheel file
- Moved `telemetry.py` to `azext_alias` folder
- Changed extension name from `azure-cli-alias-extension` to `alias`
- Added `version.py` to store the version of the extension
- Changed the URL of the extension to https://github.com/Azure/azure-cli-extensions

derekbekoe commented:
@chewong What was wrong with verify_codeowners.py?

derekbekoe commented:
@chewong Did you try not using ddt?

chewong commented:
@derekbekoe I made some changes to my [test file](https://github.com/chewong/azure-cli-alias-extension/blob/dev/azext_alias/tests/test_alias.py) without using ddt. Could you take a look and let me know what you think before moving it to this PR?

derekbekoe commented:
It looks good.
For these kinds of calls, what does 5  mean?
`alias_manager.build_collision_table(5)`
Calling the function with keyword arguments would be clearer. Also, why 5 and not 4?

derekbekoe commented on src/alias/setup.py:
```diff
@@ -0,0 +1,47 @@
+
+# Inspired by https://github.com/Azure/azure-sdk-for-python/blob/master/azure-mgmt-cdn/setup.py#L45
+extension_path = os.path.dirname(os.path.realpath(__file__))
+with open(os.path.join(extension_path, 'azext_alias/version.py'), 'r') as version_file:
```
Do not hard code the `/` here. Use os.path.join for compatibility with both Linux and Windows.
Something like `os.path.join(extension_path, 'azext_alias', 'version.py')`
